matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/worker/SessionSafeCoroutineWorker.kt

```diff
@@ -89,10 +88,10 @@ internal abstract class SessionSafeCoroutineWorker<PARAM : SessionWorkerParams>(
      * This is called when the input parameters are correct, but contain an error from the previous worker.
      */
     @CallSuper
-    open fun doOnError(params: PARAM): Result {
+    open fun doOnError(params: PARAM, failureMessage: String): Result {
         // Forward the error
-        return Result.success(inputData)
```

Comment:
previously we were always using the `inputData` which contained the `lastFailureMessage` which mean the next time the work ran the data would be restore with the pre-existing failure, causing the worker to never execute `doSafeWork`

the fix is to _consume_ the message and use the updated state for the result (so that it can be persisted without the `lastFailureMessage`)

Reply:
OK, so we do not forward the error anymore to the next worker, since it has been removed from the `params` parameter, right? (maybe update the comment above). The original idea to transmit the error was to ensure that the last worker of the chain will handle it, without breaking the whole worker chain forever.

It's still strange to me that the `lastFailureMessage` get restored when we start a new work.

Reply:
I was also surprised by this, I think it's caused by the `lastFailureMessage` causing a merge conflict in the input data

```kotlin
// requireBackgroundSync
val data = WorkerParamsFactory.toData(
        Params(
                sessionId = sessionId,
                timeout = serverTimeoutInSeconds,
                delay = 0L,
                periodic = false,
                random = UUID.randomUUID().toString()  // create a unique id for each work request
        )
)
```

Logging the work request `data` and the `inputData` on job execution 

```
// No forced crash - with random

requireBackground sync params: Data {WORKER_PARAMS_JSON : {"random":"7675"}, }
doWork input: Data {WORKER_PARAMS_JSON : {"random":"7675"}, }
requireBackground sync params: Data {WORKER_PARAMS_JSON : {"random":"107a"}, }
doWork input: Data {WORKER_PARAMS_JSON : {"random":"107a"}, }
requireBackground sync params: Data {WORKER_PARAMS_JSON : {"random":"c502"}, }
doWork input: Data {WORKER_PARAMS_JSON : {"random":"c502"}, }
requireBackground sync params: Data {WORKER_PARAMS_JSON : {"random":"a22b"}, }
doWork input: Data {WORKER_PARAMS_JSON : {"random":"a22b"}, }


// Force crash

requireBackground sync params: Data {WORKER_PARAMS_JSON : {"random":"49b1"}, }
doWork input: Data {WORKER_PARAMS_JSON : {"random":"49b1"}, }
requireBackground sync params: Data {WORKER_PARAMS_JSON : {"random":"fca1"}, }
doWork input: Data {WORKER_PARAMS_JSON : {"random":"49b1","lastFailureMessage":"error"}, }
requireBackground sync params: Data {WORKER_PARAMS_JSON : {"random":"b7e6"}, }
doWork input: Data {WORKER_PARAMS_JSON : {"random":"49b1","lastFailureMessage":"error"}, }


// Force crash removed

requireBackground sync params: Data {WORKER_PARAMS_JSON : {"random":"97c9"}, }
doWork input: Data {WORKER_PARAMS_JSON : {"random":"49b1","lastFailureMessage":"error"}, }
requireBackground sync params: Data {WORKER_PARAMS_JSON : {"random":"a2b6deb5-a33b-4361-ae3d-3dcb22913e36"}, }
doWork input: Data {WORKER_PARAMS_JSON : {"random":"49b1","lastFailureMessage":"error"}, }
```

notice that as soon as the `lastFailureMessage` is modified the request data becomes ignored, which I assume is caused by the _unspecified ordering_ of the default `OverwritingInputMerger` https://developer.android.com/reference/androidx/work/InputMerger

Reply:
> another solution is to set an InputMerger which always favours the latest input specifically for the SyncWorker

This should be far less confusing IMHO. The API of the work manager is so strange...

Reply:
(thanks for the explanations!)

Reply:
comment updated 1a21ec5f36b333ab4a02d6ed04ed49bafc931b3d

Reply:
So, do we want to use an `InputMerger` instead of consuming the error?

Reply:
I would lean towards consuming the error as having the workers stuck in the failure state sounds unintentional. Most of the workers try/catch their logic in `doSafeWork`, it only seems to be the uncaught exceptions that fall into the chain (stuck error) flow.

We have an instance of providing a merger which avoids merging for the [TimelineSendEventWorkCommon](https://github.com/vector-im/element-android/blob/develop/matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/timeline/TimelineSendEventWorkCommon.kt#L51), although the name `startChain` is a little misleading as the `NoMerger` does the opposite and disables the result chaining.

Reply:
Yes if there are chained workers we need to cancel using this mechanism, otherwise workers will be stuck if we return Failed state. This worker API is so bad...

Reply:
I think it will break some flow if we consume the failure, as we always return Success, if there is a chain of workers, the next one will be running when it shouldn't...

Reply:
in those cases, should we be using `Result.Failure` to cancel the chain? My understanding is that `APPEND_OR_REPLACE` would allow the chain to be recreated the next time a request comes in

Reply:
I'll rework the PR to make use of the `chain` extension (`NoMerge`), we can think about the worker flows as a separate ticket (as other workers may be getting stuck)

Reply:
updated 1fd1a4e82436f502b33bda2d3be8f642898689ce